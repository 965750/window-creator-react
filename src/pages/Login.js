import React, { Component } from 'react'
import { connect } from 'react-redux'
import { FormattedMessage, injectIntl } from 'react-intl'
import PropTypes from 'prop-types'
import { Redirect, Link } from 'react-router-dom'
import { login, loginRemebered } from '../store/actions/authActions'
import styles from './Login.module.scss'
import BaseCheckbox from '../components/blocks/BaseCheckbox'
import BaseInput from '../components/blocks/BaseInput'
import BaseSubmit from '../components/blocks/BaseSubmit'

class Login extends Component {
    state = {
        email: '',
        password: '',
        remember: false,
        error: null,
    }

    handleChange = e => {
        this.setState({
            [e.target.id]: e.target.value,
        })
    }

    validationError = error => {
        this.setState({
            error,
        })
    }

    toggleCheckbox = () => {
        this.setState(prevState => ({
            remember: !prevState.remember,
        }))
    }

    onLogin = e => {
        e.preventDefault()
        if (this.state.remember) {
            this.props.loginRemebered(this.state)
        } else {
            this.props.login(this.state)
        }
    }

    render() {
        if (this.props.isLoggedIn) return <Redirect to="/creator" />

        return (
            <form
                className={`${styles.form__wrapper} mt-10 mx-auto`}
                onSubmit={this.onLogin}
            >
                <BaseInput
                    value={this.state.email}
                    validationError={this.validationError}
                    handleChange={this.handleChange}
                    id="email"
                    type="text"
                    toValide={['email', 'isRequired']}
                    classes="mb-3"
                    placeholder="Email address"
                />
                <BaseInput
                    value={this.state.password}
                    validationError={this.validationError}
                    handleChange={this.handleChange}
                    id="password"
                    type="password"
                    toValide={['isRequired']}
                    classes="mb-3"
                    placeholder="Password"
                />
                <div className="flex justify-between flex-wrap mb-3">
                    <div
                        id="remember"
                        onClick={this.toggleCheckbox}
                        className="flex cursor-pointer"
                    >
                        <BaseCheckbox
                            classes="mr-2 mb-3"
                            isChecked={this.state.remember}
                        />
                        <p>
                            <FormattedMessage
                                id="Keep me logged in"
                                defaultMessage="Keep me logged in"
                            />
                        </p>
                    </div>
                    <Link
                        to="/register"
                        className="hover:text-malachite text-malachite cursor-pointer"
                    >
                        <FormattedMessage
                            id="Register new account here"
                            defaultMessage="Register new account here"
                        />
                    </Link>
                </div>
                <BaseSubmit
                    value="Login"
                    disabled={
                        !this.state.email ||
                        !this.state.password ||
                        this.state.error
                    }
                />
            </form>
        )
    }
}

const mapStateToProps = state => ({
    isLoggedIn: state.firebase.auth.uid,
})

const mapDispatchToProps = dispatch => ({
    login: credentials => dispatch(login(credentials)),
    loginRemebered: credentials => dispatch(loginRemebered(credentials)),
})

Login.propTypes = {
    isLoggedIn: PropTypes.string,
    login: PropTypes.func,
}

export default injectIntl(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )(Login)
)
