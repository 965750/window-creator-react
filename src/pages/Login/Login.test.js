import '@testing-library/jest-dom/extend-expect'

import { BrowserRouter } from 'react-router-dom'
import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react'

import Login from './index'
import messages from '../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
    stepper: {
      active: 1,
      steps: [
        {
          label: 'Step 1',
          title: 'Choose Door',
          id: 1,
        },
        {
          label: 'Step 2',
          title: 'Choose Door Division',
          id: 2,
        },
        {
          label: 'Step 3',
          title: 'Choose Color',
          id: 3,
        },
      ],
    }
  },
  creator: {
    window: {
      color: '#000',
      width: 110,
      height: 230,
      doorType: 2,
      rows: [0, 1, 2],
      columns: [0, 1, 2],
    },
  },
  firebase: {
    auth: {
      uid: null,
    }
  }
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    case 'MOCK_LOGIN':
      return {
        ...state,
        firebase: {
          auth: {
            uid: 'mockUid',
          }
        }
      }
    case 'RESET_STORE':
      return initialStore
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('Login', () => {
  let wrapper
  const initialProps = {
    window: {
      color: '#000',
      width: 110,
      height: 230,
      doorType: 2,
      rows: [0, 1, 2],
      columns: [0, 1, 2],
    },
    resizeWindow: jest.fn(),
    changeWindowType: jest.fn(),
    setNotification: jest.fn(),
  }

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <BrowserRouter>
        <IntlProvider
          locale="en"
          messages={messages[store.getState().theme.lang]}
        >
          <Login {...props} />
        </IntlProvider>
      </BrowserRouter>,
    )
  }

  afterEach(() => {
    cleanup

    store.dispatch({
      type: 'RESET_STORE',
    })
  })

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should redirect when user is logged in', () => {
    setUp()

    store.dispatch({
      type: 'MOCK_LOGIN',
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should show checkbox checked on click', () => {
    setUp()

    fireEvent.click(wrapper.getByTestId('checkboxWrapper'))

    expect(wrapper.getByTestId('checkboxIcon')).not.toBeNull()
  })

  it('should enable submit button after correctly filling form', () => {
    setUp()

    expect(wrapper.getByTestId('submitWrapper').disabled).toBe(true)

    fireEvent.change(wrapper.getByPlaceholderText('Adres Email'), { target: { value: 'mockedEmail@gmail.com' }})
    fireEvent.change(wrapper.getByPlaceholderText('Has≈Ço'), { target: { value: 'mockPassword' }})

    expect(wrapper.getByTestId('submitWrapper').disabled).toBe(false)
  })
})
