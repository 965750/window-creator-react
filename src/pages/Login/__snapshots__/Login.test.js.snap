// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="form__wrapper mt-10 mx-auto"
  >
    <div
      class="field__wrapper mb-3"
      data-testid="inputWrapper"
    >
      <input
        class="field  border border-silver rounded p-3 w-full"
        id="email"
        placeholder="Adres Email"
        type="text"
        value=""
      />
    </div>
    <div
      class="field__wrapper mb-3"
      data-testid="inputWrapper"
    >
      <input
        class="field  border border-silver rounded p-3 w-full"
        id="password"
        placeholder="Hasło"
        type="password"
        value=""
      />
    </div>
    <div
      class="flex justify-between flex-wrap mb-3"
    >
      <div
        class="flex cursor-pointer"
        id="remember"
      >
        <div
          class="checkbox__wrapper border-gullGray border relative mr-2 mb-3"
          data-testid="checkboxWrapper"
        />
        <p>
          Pozostań zalogowany
        </p>
      </div>
      <a
        class="hover:text-malachite text-malachite cursor-pointer"
        href="/register"
      >
        Stwórz nowe konto tutaj
      </a>
    </div>
    <input
      class="submit  mx-auto cursor-pointer block bg-mountainMeadow rounded text-white"
      data-testid="submitWrapper"
      disabled=""
      type="submit"
      value="Zaloguj"
    />
  </form>
</DocumentFragment>
`;

exports[`Login should redirect when user is logged in 1`] = `<DocumentFragment />`;
